src/Microsoft.OData.Core/ODataBatchReaderStream.cs

```diff
@@ -343,308 +80,58 @@ internal int ReadWithDelimiter(byte[] userBuffer, int userBufferOffset, int coun
-                    //       because the batch reader will skip over everything (incl. the preamble)
-                    //       until it finds the first changeset (or batch) boundary
-                    this.changesetEncoding = this.DetectEncoding();
-                }
-
-                // Verify that we only allow single byte encodings and UTF-8 for now.
-                ReaderValidationUtils.ValidateEncodingSupportedInBatch(this.changesetEncoding);
-            }
-            else if (this.ChangeSetBoundary != null)
-            {
-                headers.TryGetValue(ODataConstants.ContentIdHeader, out contentId);
-            }
-
-            return isChangeSetPart;
-        }
-
-        /// <summary>
-        /// Reads the headers of a batch part or an operation.
-        /// </summary>
-        /// <returns>A dictionary of header names to header values; never null.</returns>
-        internal ODataBatchOperationHeaders ReadHeaders()
-        {
-            Debug.Assert(this.batchEncoding != null, "Batch encoding should have been established on first call to SkipToBoundary.");
-
-            // [Astoria-ODataLib-Integration] WCF DS Batch reader compares header names in batch part using case insensitive comparison, ODataLib is case sensitive
-            ODataBatchOperationHeaders headers = new ODataBatchOperationHeaders();
-
-            // Read all the headers
-            string headerLine = this.ReadLine();
-            while (headerLine != null && headerLine.Length > 0)
-            {
-                string headerName, headerValue;
-                ValidateHeaderLine(headerLine, out headerName, out headerValue);
-
-                if (headers.ContainsKeyOrdinal(headerName))
-                {
-                    throw new ODataException(Strings.ODataBatchReaderStream_DuplicateHeaderFound(headerName));
-                }
-
-                headers.Add(headerName, headerValue);
-                headerLine = this.ReadLine();
-            }
-
-            return headers;
-        }
-
-        /// <summary>
-        /// Read and return the next line from the batch stream, skipping all empty lines.
-        /// </summary>
-        /// <remarks>This method will throw if end-of-input was reached while looking for the next line.</remarks>
-        /// <returns>The text of the first non-empty line (not including any terminating newline characters).</returns>
-        internal string ReadFirstNonEmptyLine()
+        /// <remarks>
+        /// Default implementation at base level is no-op.
+        /// Derived types provide specific implementation as needed.
+        /// Note that, instead of defining IDisposable on the root level and causing ripples to handle IDisposable
+        /// in other related types, we use this out-of-band method and its overrides for disposing.
+        /// </remarks>
+        protected internal virtual void DisposeResources()

```

Comment:
>protected internal virtual void DisposeResources() [](start = 8, length = 50)

This doesn't get called from anywhere, and isn't implemented in the MultipartMixedBatchReaderStream.  If it's only implemented in the JSONBatchReaderStream, and (presumably) only called from the JSONBatchReader, does it need to be in the base class?

Reply:
It is overridden in ODataJsonLightBatchBodyContentReaderStream, see on 6.x branch: [https://github.com/biaol-odata/odata.net/blob/maintenance-6.x_JsonFormatBatch-A/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchBodyContentReaderStream.cs#L267](https://github.com/biaol-odata/odata.net/blob/maintenance-6.x_JsonFormatBatch-A/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchBodyContentReaderStream.cs#L267)

---
In reply to: [141114159](https://github.com/OData/odata.net/pull/960#discussion_r141114159) [](ancestors = 141114159)

Reply:
ok; makes sense.

---
In reply to: [142294609](https://github.com/OData/odata.net/pull/960#discussion_r142294609) [](ancestors = 142294609,141114159)